mv: fail whole requests instead of dropping view updates when overloaded

wmitros opened the pull request:
Since 4c767c379cc61de3322e4874c33481b11d31e671 we can reach a situation where we know that we have admitted too many expensive view update operations and the mechanism of dropping the following view updates can be triggerred. Instead, to avoid creating inconsistencies between base replicas and views, we want to reject entire requests that we can't fully handle.
This patch changes the older behavior to the new one.

nyh commented:
The PR's title says "fail whole requests" but (please correct me if I'm wrong) it's not what I see here: You don't fail the user's "request" - which you'd need to do in the coordinator. Instead you fail just one (base-) replica write. The code comment you quoted in https://github.com/scylladb/scylladb/issues/17426 (written by Duarte some six years ago) suggested that this can lead to unwelcome inconsistencies between different base-table replicas - which aren't terrible (because repair fixes them) but aren't great either - which is why this wasn't done originally.

Another problem with this solution is that it fails a request (or part of it) after already doing quite a bit of work for it. Since the purpose of admission control is to reduce the load, it's better to drop the request before doing work on it - not after.

What I suggested to do in https://github.com/scylladb/scylladb/issues/17426, instead of failing the replica write, is to fail the coordinator write before starting on it. The coordinator has an estimate of the backlog of the replicas (which it gets from these replicas), and can decide the replicas are two overloaded with view updates and refuse immediately - without doing any work - any request which involves view updates.

nyh commented:
That being said, I think that if the title of the PR is changed a bit, it's a good PR and should be merged. What your patch actually does is to prefer creating inconsistencies between base replicas, instead of creating an inconsistency between a base replica and a view replica which the current code does. The rationale is that we already have a good way to fix inconsistencies between base replicas (i.e., regular repair) - but don't have a good way to fix base-view inconsistencies.

Please create an issue that this PR will "Fixes". I thought we had one (we discussed this code in the past, and I believe it was on Eliran's hitlist), but I can't find any.

wmitros commented:
I added the issue that this `Fixes` and changed the commit message accordingly. Also, I reused most of the former comment that we had here as we still want to improve this in the future with actual admission control.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7388/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7388/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/478/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8511/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7388/)
#### Build Details:
- Duration: 3 hr 48 min
- Builder: spider4.cloudius-systems.com

nyh commented:
@wmitros please say if your decision to add backport/none was based on something. On one hand, this is a consistency bug fix should should be backported. On the other hand, because nobody - not you nor QA - actually plans to **test** this change and its implications, maybe we should consider the backport not safe enough?

Procedural comment (@mykaul @yaronkaikov): As I feared, when a developer sends a PR, instead of missing a backport tag, they are now "forced" to add a `backport/none` tag. For me as a reviewer and maintainer this tag is worse than nothing - it doesn't tell me if the developer thought hard about backports and decided that a backport isn't needed, or just took the path of least resistance. I can pretend everything is fine, and since the developer said no backport is needed, then no backport is needed. But is that true?

nyh commented:
Merged. Let's hope this patch will be as lucky as its number :-)

We should continue discussing the backports we should or shouldn't do when the patch is promoted - justify why it isn't needed, or decide that it is needed.

wmitros commented:
> @wmitros please say if your decision to add backport/none was based on something. On one hand, this is a consistency bug fix should should be backported. On the other hand, because nobody - not you nor QA - actually plans to **test** this change and its implications, maybe we should consider the backport not safe enough?

I was also thinking that there are arguments both for and against backporting and decided not to because you could argue this isn't a bug fix, just a change in direction and perhaps some users would find dropping view updates to be the correct behavior.

yaronkaikov commented:
> > I think it's a step forward than the bot setting unconditionally 'backport candidate' to an issue, no?
> 
> ‎That's not the way I see it... What I fear is that all developers will just use "backport/none" on all new PRs, and then we won't have any backports and won't even have a reminding that we need to consider backporting (which is what that "backport candidate" solve). Maybe we can solving by enforcing a rule that a patch will not be merged without a textual discussion (not tags!) of which backports are needed and why. But as long as we merge PRs with backport/none without demanding such discussion (like I just did now in this PR, because the PR itself is good), we'll never remember to revisit this decision.

@nyh , I think we need some time so people will adopt the new process. One of the reasons to move forward with this project is the fact that having a `Backport candidate` label on promoted issues didn't make our backport process/decision fast enough. 
Today we have both, at least for the next few weeks 

Maybe a textual discussion is a good idea, so why not do it in case someone added a `backport/none` label?

mykaul commented:
> > Maybe a textual discussion is a good idea, so why not do it in case someone added a `backport/none` label?
> 
> Yes, as we saw here, we do need reviewers to ask the developer to explain why he chose to use backport/none if it's not obvious. But the problem is more subtle: Let's say I'm not convinced he's right. Maybe I (as a reviewer or maintainer) need to think about it some more. Maybe I think we need to ask Tzach or Avi or Yaniv for their opinions on whether this should be backported or not. Do I not merge the PR until this decision is made? And if I do merge it, because the PR itself is good (like I did in this PR!), how will I remember a week from now that we forgot to make the decision about the packport of this PR? The need to not forget to make that decision was why the "backport candidate" tag was invented. I'm not sure that it can go away now or in a few weeks.

Merging has little to do with deciding on backport or not (I think?).
Perhaps we do not need to enforce 'backport/none' , but removing 'backport candidate' ideally should also be accompanied by an explanation - which you may or may not be convinced of.

yaronkaikov commented:
> > Maybe a textual discussion is a good idea, so why not do it in case someone added a `backport/none` label?
> 
> Yes, as we saw here, we do need reviewers to ask the developer to explain why he chose to use backport/none if it's not obvious. But the problem is more subtle: Let's say I'm not convinced he's right. Maybe I (as a reviewer or maintainer) need to think about it some more. Maybe we need to ask Tzach Avi or Yaniv for their opinions on whether this should be backported. Do I not merge the PR until this decision is made? And if I do merge it because the PR itself is good (like I did in this PR!), how will I remember a week from now that we forgot to decide the backport of this PR? The need to not forget to make that decision was why the "backport candidate" tag was invented. I'm not sure that it can go away now or in a few weeks.

I understand. So maybe the solution should be either choose `backport/x.y` or `Backport candidate` from the start? 
The difference from today is if we realize in a few weeks that one of the issues/PR wasn't backported, we can just change the label, and the magic will happen.

nyh commented:
> Merging has little to do with deciding on backport or not (I think?)

Exactly. Which was the idea of the old "backport candidate" tag - you merge now, and when the merge is over the backport candidate tag is added to it to remind you that someone needs to make a final decision regarding the backport. The talk now is about **removing** this tag (after a few week adjustment period), and I'm asking the question if that's a good idea, and if we remove it, how will we remember to go back to make this decision.

If the PR had backport/5.4, the automated process will create a separate PR and we won't forget it. But if the PR had backport/none (as in this case), nothing will remind us. Arguably, even worse, if the PR author added just backport/5.4 and not backport/5.2, for some undiscussed and unjustified reason, nobody will remember to discuss this reason because a PR for the 5.4 backport will exist and when somebody goes to consider it, he won't even know that the 5.2 backport was never created.

> Perhaps we do not need to enforce 'backport/none' , but removing 'backport candidate' ideally should also be accompanied by an explanation - which you may or may not be convinced of.

When I talked about "removing backport candidate" I talked about the planned (if I understood correctly) removal of the "backport candidate" bot.

But if you're talking about the different subject of manually removing the backport candidate tag, you're right: Until now one of the maintainers removed this tag later when considering backports, always accompaying this change with a textual comment. More than once, the decision was wrong (the maintainer explained why we shouldn't backport it it, and a year later somebody encountered this bug and we regretted not backporting) - but it was at least a decision with some justifications and not some random unexplained decision.
